### PR TITLE
mattdrayer/ECOM-4683: Return only child seat products in API response

### PR DESCRIPTION
**Pull request by mattdrayer**
@mjfrey @bderusha -- Enrollment Code products were being included in the serialization workflow for the Course Administration Tool course product list.  I've added a filter to return only "child" products for the course (Enrollment Codes are "standalone").

### CODE REVIEW

**Comment by clintonb** [comment]
There is value to including all products, for example if/when we want to include a link to the enrollment code product on the marketing site. This filtering should be done by the Course Administration Tool.

**Comment by mattdrayer** [comment]
As I understand it, the intent isn't to list the enrollment code directly -- the bulk purchasing variation of any given seat product is indicated by the 'bulk_sku' attribute, which may or may not contain a value.

**Comment by mattdrayer** [comment]
This pull request depends on https://github.com/edx/ecommerce/pull/786

**Comment by mattdrayer** [comment]
Would love to get through this PR today -- @mjfrey @bderusha @clintonb --  do you guys have anything you'd like to see changed or can I move forward?

**Comment by mjfrey** [comment]
LGTM:
👍

**Comment by clintonb** [comment]
This introduces a backwards-incompatible change to the API that necessitates a major version bump. Instead of changing the API, change the client code. Update CAT to only look at course seat child products. Otherwise, we will run into this issue the next time we add a course-related product.

**Comment by mattdrayer** [comment]
I'm not sure I understand why this is being viewed as a backwards-incompatible change -- the API endpoint has not changed, nor the data schema.  In retrospect, the incompatible change was actually the addition of the new Enrollment Code product class, which was not expected by the CAT client and resulted the misbehavior of that tool.  So with this change set we are actually repairing the original semantics of the 'course product list' API operation by introducing this filtering.

File: ecommerce/static/js/models/course_model.js [hunk 1]
```diff
@@ -135,6 +135,7 @@ define([
             initialize: function () {
                 this.get('products').on('change:id_verification_required', this.triggerIdVerified, this);
                 this.on('sync', this.removeParentProducts, this);
+                this.on('sync', this.removeStandaloneProducts, this);
```
Comment (mattdrayer) [inline ecommerce/static/js/models/course_model.js]: @clintonb -- here's one approach from the client/CAT side that solves the issue -- let me know if it's acceptable and I'll remove the server-side solution from this PR.

Comment (clintonb) [inline ecommerce/static/js/models/course_model.js]: The approach is correct; however, you should be removing products where `product_class` is not equal to `Seat`.

**Comment by clintonb** [comment]
The server-side change is backwards-incompatible in the sense that it removes data that was previously being provided by the API. We intentionally opted not to limit the data returned in the `products` array to avoid later confusion about what is/is not included.

The client-side change is much more desirable. You just need to update it to filter on product class.

File: ecommerce/static/js/models/course_model.js [hunk 2]
```diff
@@ -184,18 +184,52 @@ define([
...
+                var parents = products.where({structure: 'parent'});
                 products.remove(parents);
             },
 
             /**
+             * Removes "standalone" products from the Product collection.
+             *
+             * This product is never exposed to the user, and should be ignored for all data operations.
+             */
+            removeStandaloneProducts: function () {
```
Comment (clintonb) [inline ecommerce/static/js/models/course_model.js]: Is this necessary? I'm not aware of any standalone seat products.

Comment (mattdrayer) [inline ecommerce/static/js/models/course_model.js]: I guess it's not at this point -- "Enrollment Code" product_class products have 'standalone' structure -- I was thinking that maybe the scenario would arise at some point, but no sense having this hanging around to cause problems if it's not needed.

File: ecommerce/extensions/api/v2/tests/views/test_courses.py [hunk 3]
```diff
@@ -49,7 +49,7 @@ def serialize_course(self, course, include_products=False):
         }
 
         if include_products:
-            data['products'] = [self.serialize_product(product) for product in course.products.all()]
+            data['products'] = [self.serialize_product(product) for product in course.seat_products]
```
Comment (mattdrayer) [inline ecommerce/extensions/api/v2/tests/views/test_courses.py]: This is an artifact from the server-side attempt, I'll remove shortly

**Comment by clintonb** [comment]
👍 (once commits squashed, and comments addressed)

Note that you'll need DevOps to deploy to stage as there are migrations in the latest batch of commits. Feel free to rebase the tag `release-2016-06-14`.

**Comment by mattdrayer** [comment]
Cool -- PR is cleaned up now -- will merge and update the tag after the build goes green

**Pull request merged**